rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,178 @@
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // Return early unless it's one of the sends we are interested in
+        if (send->fun != core::Names::abstract() && send->fun != core::Names::final_() &&
+            send->fun != core::Names::override_() && send->fun != core::Names::overridable()) {
+            return tree;
+        }
+
+        // This is still an invocation where `sig` is the first receiver, but there is still no block. We need to pop
+        // the last entry for incomplete sigs and add the current send to account for multiple methods chained on sig
+        //
+        // E.g.: `sig.override.final {}`
+        //
+        // If we never find a send where the first receiver is a `sig` and the last invocation has the declaration
+        // block, then we add the error later on
+        if (!send->hasBlock()) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Signature declarations expect a block");
+                e.addErrorNote("Complete the signature by adding a block declaration: sig `{}`", "{ ... }");
+            }
+
+            return tree;
+        }
+
+        // For all other cases, we have to re-write the block
+        // E.g.: `sig.abstract { void }` -> `sig { abstract.void }`
+        auto *block = send->block();
+        auto blockBody = ast::cast_tree<ast::Send>(block->body);
+
+        // If the blockBody is not a send, then we have a sequence of instructions inside the signature block
+        if (blockBody == nullptr) {
+            if (auto e = ctx.beginError(block->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Malformed signature: cannot have multiple instructions inside a signature block");
+            }
+
+            return tree;
+        }
+
+        return buildReplacement(ctx, send, blockBody);
+    }
+
+    bool firstReceiverIsSig(ast::Send *send) {
+        do {
+            if (send->fun == core::Names::sig()) {
+                this->previousSigSend = send;
+                return true;
+            }
+            send = ast::cast_tree<ast::Send>(send->recv);
+        } while (send);
+
+        return false;
+    }
+
+    void checkDuplicates(core::MutableContext ctx, ast::Send *sigBlock, core::NameRef fun) {
+        while (sigBlock) {
+            if (sigBlock->fun == fun) {
+                if (auto e = ctx.beginError(sigBlock->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Duplicate invocation of `{}` in signature declaration", fun.toString(ctx));
+                }
```

Comment:
I think we can give a better location for the error here, since `sigBlock` has the location for the actual function name.  We can also provide in a note the original location for the method name.